### PR TITLE
feat: support getting Kubernetes resources through gRPC gateway

### PR DESCRIPTION
**Pull request by Unix4ever**
Fixes: https://github.com/talos-systems/theila/issues/17
    
Generate gRPC gateway Go and TypeScript definitions and pull in Talos
protobufs to reuse some messages from there.
    
Rewampe API:
- make `Get` return `interface{}` instead of writing the response
into `dest` parameter of the function.
- split `Get` and `List` API into two functions.
    
Set up gRPC handler with `/api` prefix.
    
Create resource gRPC service which is similar to Talos but supports
several clusters and different sources (Talos/Kubernetes).
  
Also:
- added selectors support to the watch.
- use Talos `WatchRequest` instead of plain `resource` string to support
getting resource from a specific namespace.

From this point we should have all kinds of concepts in place to query anything we might want from the cluster defined in the Kubeconfig, or from the cluster created by Sidero

Signed-off-by: Artem Chernyshev <artem.0xD2@gmail.com>

### CODE REVIEW

File: api/rpc/resource.proto [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+  common.Source source = 3;
+}
+
+message GetFromClusterResponse {
+  // Body should contain JSON encoded spec.
+  string body = 1;
+}
+
+message ListFromClusterRequest {
+  resource.ListRequest resource = 1;
```
Comment (Unix4ever) [inline api/rpc/resource.proto:23]: Embed protobuf messages from Talos here.

File: api/rpc/resource.proto [hunk 2]
```diff
@@ -0,0 +1,40 @@
+syntax = "proto3";
+package resource;
+
+option go_package = "github.com/talos-systems/theila/api/rpc";
+
+import "talos/resource/resource.proto";
+import "common/theila.proto";
+
+message GetFromClusterRequest {
+  resource.GetRequest resource = 1;
```
Comment (Unix4ever) [inline api/rpc/resource.proto:10]: Embed protobuf messages from Talos here.

File: api/socket/message.proto [hunk 3]
```diff
@@ -38,31 +32,13 @@ message Message {
...
-  // Cluster fetches the context from the cluster resource in Kubernetes.
-  Cluster cluster = 2;
-}
-
 message WatchSpec {
   // Resource name to watch.
   string resource = 1;
   // Source to get the watch data from.
-  Source source = 2;
+  common.Source source = 2;
```
Comment (Unix4ever) [inline api/socket/message.proto:39]: This is now shared between grpc calls and ws messages.

File: frontend/src/api/resource.ts [hunk 4]
```diff
@@ -0,0 +1,459 @@
+/* eslint-disable */
+import { util, configure, Reader, Writer } from "protobufjs/minimal";
```
Comment (Unix4ever) [inline frontend/src/api/resource.ts:2]: Here we generate protobufs for typescript. And as it has services it won't work in the web.
I think I tried to compile it before and it fails.

But as we use protobuf messages for websockets I kept that enabled for all `.proto` files.
There is option to add one more flag that will enable either gateway mode or protobuf mode for each `.proto` file, but I didn't like how it looked in the code, but can bring that back.

File: frontend/src/talos/resource/resource.pb.ts [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+}
+
+export type WatchResponse = {
+  metadata?: CommonCommon.Metadata
+  eventType?: EventType
+  definition?: Resource
+  resource?: Resource
+}
+
+export class ResourceService {
```
Comment (Unix4ever) [inline frontend/src/talos/resource/resource.pb.ts:73]: That's the main thing we are going to use in Typescript code: generated gateway client.

File: internal/backend/runtime/kubernetes/kubernetes.go [hunk 6]
```diff
@@ -122,25 +118,57 @@ func (r *Runtime) Get(ctx context.Context, dest interface{}, setters ...runtime.
...
+
+// List implements runtime.Runtime.
+func (r *Runtime) List(ctx context.Context, setters ...runtime.QueryOption) (interface{}, error) {
+	opts := runtime.NewQueryOptions(setters...)
+
+	// turn single item into a list
+	if opts.Resource != "" {
+		parts := strings.Split(opts.Resource, ".")
+		if !strings.HasSuffix(parts[0], "List") {
+			parts[0] += "List"
```
Comment (Unix4ever) [inline internal/backend/runtime/kubernetes/kubernetes.go:139]: Maybe there's a better way, but gosh... Kubernetes clients are driving me nuts. Need to retry looking for that with a fresh head.

File: frontend/src/views/Demo.vue [hunk 7]
```diff
@@ -50,6 +55,30 @@ import {
         "Talos",
       ],
     };
+  },
+
+  methods: {
+    async fetch() {
+      console.log(await ClusterResourceService.Get({
```
Comment (Unix4ever) [inline frontend/src/views/Demo.vue:62]: Fast and dirty demo of how it's going to be used in the frontend.

Probably need export `pathPrefix` settings to some global variable.
As we mount grpc at `/api` path we have to use this prefix everywhere.

Additionally, I don't 100% like that in Watch I use GVR (GroupVersionResource) and here I only figured out how to get the resource using GVK (GroupVersionKind) and again I couldn't find a way to convert Resource to Kind yet.

File: api/rpc/resource.proto [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+  common.Context context = 2;
+  // Data source to use to get the resource.
+  common.Source source = 3;
+  // Selectors allow filtering list results by labels.
+  repeated string selectors = 4;
+}
+
+message ListFromClusterResponse {
+  // Body should contain JSON encoded list spec.
+  string body = 1;
```
Comment (Unix4ever) [inline api/rpc/resource.proto:34]: Make repeated.

**Comment by Unix4ever** [comment]
All to get this for now: https://github.com/talos-systems/theila/pull/21/files#diff-125d8faacce1d9809754f7ffd7e43fbeab941075d817adc1a202cf0f32502a20R93

**Review by smira — APPROVED** [review]

**Pull request merged**